OPAL/ai/src/it/resources/org/opalj/fpcf/analyses/SimplePurityTest-lambdas-1.8-g-parameters-genericsignature.jar.txt
DefinedMethod(ObjectType(annotations/target/CallGraphAlgorithm),<clinit>,MethodDescriptor((): void),annotations.target.CallGraphAlgorithm{ static void <clinit>() }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/CallGraphAlgorithm),<init>,MethodDescriptor((java.lang.String, int): void),annotations.target.CallGraphAlgorithm{ private void <init>(java.lang.String,int) }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/CallGraphAlgorithm),valueOf,MethodDescriptor((java.lang.String): annotations.target.CallGraphAlgorithm),annotations.target.CallGraphAlgorithm{ public static annotations.target.CallGraphAlgorithm valueOf(java.lang.String) }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/CallGraphAlgorithm),values,MethodDescriptor((): annotations.target.CallGraphAlgorithm[]),annotations.target.CallGraphAlgorithm{ public static annotations.target.CallGraphAlgorithm[] values() }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/TargetResolution),<clinit>,MethodDescriptor((): void),annotations.target.TargetResolution{ static void <clinit>() }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/TargetResolution),<init>,MethodDescriptor((java.lang.String, int): void),annotations.target.TargetResolution{ private void <init>(java.lang.String,int) }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/TargetResolution),valueOf,MethodDescriptor((java.lang.String): annotations.target.TargetResolution),annotations.target.TargetResolution{ public static annotations.target.TargetResolution valueOf(java.lang.String) }) => ImpureByAnalysis
DefinedMethod(ObjectType(annotations/target/TargetResolution),values,MethodDescriptor((): annotations.target.TargetResolution[]),annotations.target.TargetResolution{ public static annotations.target.TargetResolution[] values() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/EitherOrPredicate),<init>,MethodDescriptor((java.util.function.Predicate, java.util.function.Predicate): void),lambdas.EitherOrPredicate{ public void <init>(java.util.function.Predicate,java.util.function.Predicate) }) => Pure
DefinedMethod(ObjectType(lambdas/EitherOrPredicate),lambda$0,MethodDescriptor((java.lang.String): boolean),lambdas.EitherOrPredicate{ static boolean lambda$0(java.lang.String) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/EitherOrPredicate),main,MethodDescriptor((java.lang.String[]): void),lambdas.EitherOrPredicate{ public static void main(java.lang.String[]) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/EitherOrPredicate),test,MethodDescriptor((java.lang.Object): boolean),lambdas.EitherOrPredicate{ public boolean test(java.lang.Object) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),<init>,MethodDescriptor((): void),lambdas.HigherOrder{ public void <init>() }) => Pure
DefinedMethod(ObjectType(lambdas/HigherOrder),higherOrderPredicate,MethodDescriptor((): void),lambdas.HigherOrder{ public void higherOrderPredicate() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),lambda$0,MethodDescriptor((java.util.function.Predicate): boolean),lambdas.HigherOrder{ static boolean lambda$0(java.util.function.Predicate) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),lambda$1,MethodDescriptor((java.lang.String): boolean),lambdas.HigherOrder{ static boolean lambda$1(java.lang.String) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),lambda$2,MethodDescriptor((java.lang.String): boolean),lambdas.HigherOrder{ static boolean lambda$2(java.lang.String) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),lambda$3,MethodDescriptor((java.lang.Character): boolean),lambdas.HigherOrder{ static boolean lambda$3(java.lang.Character) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/HigherOrder),nestedPredicate,MethodDescriptor((): void),lambdas.HigherOrder{ public void nestedPredicate() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InstanceCapturing),$forward$lambda$0,MethodDescriptor((): void),lambdas.InstanceCapturing{ void $forward$lambda$0() }) => Pure
DefinedMethod(ObjectType(lambdas/InstanceCapturing),$forward$lambda$1,MethodDescriptor((): void),lambdas.InstanceCapturing{ void $forward$lambda$1() }) => Pure
DefinedMethod(ObjectType(lambdas/InstanceCapturing),$forward$lambda$2,MethodDescriptor((): void),lambdas.InstanceCapturing{ void $forward$lambda$2() }) => Pure
DefinedMethod(ObjectType(lambdas/InstanceCapturing),<init>,MethodDescriptor((): void),lambdas.InstanceCapturing{ public void <init>() }) => Pure
DefinedMethod(ObjectType(lambdas/InstanceCapturing),captureArray,MethodDescriptor((): void),lambdas.InstanceCapturing{ public void captureArray() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InstanceCapturing),captureObject,MethodDescriptor((): void),lambdas.InstanceCapturing{ public void captureObject() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InstanceCapturing),capturePrimitive,MethodDescriptor((): void),lambdas.InstanceCapturing{ public void capturePrimitive() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InstanceCapturing),lambda$0,MethodDescriptor((): void),lambdas.InstanceCapturing{ private void lambda$0() }) => DSideEffectFree
DefinedMethod(ObjectType(lambdas/InstanceCapturing),lambda$1,MethodDescriptor((): void),lambdas.InstanceCapturing{ private void lambda$1() }) => DSideEffectFree
DefinedMethod(ObjectType(lambdas/InstanceCapturing),lambda$2,MethodDescriptor((): void),lambdas.InstanceCapturing{ private void lambda$2() }) => DSideEffectFree
DefinedMethod(ObjectType(lambdas/InvokeDynamics),<init>,MethodDescriptor((): void),lambdas.InvokeDynamics{ public void <init>() }) => Pure
DefinedMethod(ObjectType(lambdas/InvokeDynamics),lambda$0,MethodDescriptor((int, int): java.lang.Integer),lambdas.InvokeDynamics{ static java.lang.Integer lambda$0(int,int) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InvokeDynamics),main,MethodDescriptor((java.lang.String[]): void),lambdas.InvokeDynamics{ public static void main(java.lang.String[]) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/InvokeDynamics),simpleLambdaAdd,MethodDescriptor((int, int): int),lambdas.InvokeDynamics{ public int simpleLambdaAdd(int,int) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),$forward$lambda$2,MethodDescriptor((): void),lambdas.Lambdas{ void $forward$lambda$2() }) => Pure
DefinedMethod(ObjectType(lambdas/Lambdas),$forward$lambda$3,MethodDescriptor((int): void),lambdas.Lambdas{ void $forward$lambda$3(int) }) => Pure
DefinedMethod(ObjectType(lambdas/Lambdas),<init>,MethodDescriptor((): void),lambdas.Lambdas{ public void <init>() }) => Pure
DefinedMethod(ObjectType(lambdas/Lambdas),complexBodyWithReturns,MethodDescriptor((): java.lang.Integer),lambdas.Lambdas{ public java.lang.Integer complexBodyWithReturns() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),doubleDoubleParamter,MethodDescriptor((): java.lang.Double),lambdas.Lambdas{ public java.lang.Double doubleDoubleParamter() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),doubleFloatParamter,MethodDescriptor((): java.lang.Double),lambdas.Lambdas{ public java.lang.Double doubleFloatParamter() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),doubleFloatParamter2,MethodDescriptor((): double),lambdas.Lambdas{ public double doubleFloatParamter2() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),doubleIntParamter,MethodDescriptor((): java.lang.Double),lambdas.Lambdas{ public java.lang.Double doubleIntParamter() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),instanceClosure,MethodDescriptor((): void),lambdas.Lambdas{ public void instanceClosure() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),intDoubleParamter,MethodDescriptor((): java.lang.Double),lambdas.Lambdas{ public java.lang.Double intDoubleParamter() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),intLongParamter,MethodDescriptor((): java.lang.Long),lambdas.Lambdas{ public java.lang.Long intLongParamter() }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$0,MethodDescriptor((): void),lambdas.Lambdas{ static void lambda$0() }) => DPure
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$1,MethodDescriptor((int): void),lambdas.Lambdas{ static void lambda$1(int) }) => DPure
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$10,MethodDescriptor((java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$10(java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$11,MethodDescriptor((java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$11(java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$12,MethodDescriptor((java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$12(java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$13,MethodDescriptor((java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$13(java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$14,MethodDescriptor((java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$14(java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$15,MethodDescriptor((java.lang.Integer, java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$15(java.lang.Integer,java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$16,MethodDescriptor((java.lang.Integer, java.lang.Integer): java.lang.Integer),lambdas.Lambdas{ static java.lang.Integer lambda$16(java.lang.Integer,java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$17,MethodDescriptor((java.lang.Double): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$17(java.lang.Double) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$18,MethodDescriptor((java.lang.Double, java.lang.Double): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$18(java.lang.Double,java.lang.Double) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$19,MethodDescriptor((java.lang.Double, java.lang.Integer): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$19(java.lang.Double,java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$2,MethodDescriptor((): void),lambdas.Lambdas{ private void lambda$2() }) => DSideEffectFree
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$20,MethodDescriptor((java.lang.Integer, java.lang.Double): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$20(java.lang.Integer,java.lang.Double) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$21,MethodDescriptor((java.lang.Double, java.lang.Float): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$21(java.lang.Double,java.lang.Float) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$22,MethodDescriptor((double, float): double),lambdas.Lambdas{ static double lambda$22(double,float) }) => Pure
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$23,MethodDescriptor((java.lang.Long): java.lang.Long),lambdas.Lambdas{ static java.lang.Long lambda$23(java.lang.Long) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$24,MethodDescriptor((java.lang.Long, java.lang.Long): java.lang.Long),lambdas.Lambdas{ static java.lang.Long lambda$24(java.lang.Long,java.lang.Long) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$25,MethodDescriptor((java.lang.Long, java.lang.Integer): java.lang.Long),lambdas.Lambdas{ static java.lang.Long lambda$25(java.lang.Long,java.lang.Integer) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$26,MethodDescriptor((java.lang.Integer, java.lang.Long): java.lang.Long),lambdas.Lambdas{ static java.lang.Long lambda$26(java.lang.Integer,java.lang.Long) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$27,MethodDescriptor((java.lang.Long, java.lang.Float): java.lang.Float),lambdas.Lambdas{ static java.lang.Float lambda$27(java.lang.Long,java.lang.Float) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$28,MethodDescriptor((java.lang.Long, java.lang.Double): java.lang.Double),lambdas.Lambdas{ static java.lang.Double lambda$28(java.lang.Long,java.lang.Double) }) => ImpureByAnalysis
DefinedMethod(ObjectType(lambdas/Lambdas),lambda$29,MethodDescriptor((java.util.ArrayList, java.util.HashMap, java.util.HashMap): void),lambdas.Lambdas{ static void lambda$29(java.util.ArrayList,java.util.HashMap,java.util.HashMap) }) => ImpureByAnalysis

Changes.md
# Major Changes

## Upcoming
 - Added support for instrumenting class files
 - support for Java 9
 - Hermes now has extended visualization capabilities to make it even easiere to comprehend the differences between projects
 - The overall performance has been improved (in particular on multi-core systems with 4 or more cores)
 - Moved to sbt 1.1.x
 - Fixed issues in some test which open a huge number of files
 - Fixed a rare issue in the identification of closed strongly connected components


## 1.0.0 - Released Oct. 25th 2017
 - *we are now using Scala 2.12.4*
 - added a method to reset a Project to its initial state (all information derived by analyses is thrown away)
 - added several type test/cast methods to Instruction to support cases where the type of an instruction is known from the context
 - added a new framework for testing properties derived using the `PropertiesStore`
 - improved the time required to create a project by ~20-30%
 - improved handling of exceptions in TACAI
 - improved the framework for computing control-dependency information
 - improved the precision and soundness of the FieldMutabilityAnalysis
 - fixed an off-by-one-error in computeMaxLocals
 - fixed the recording of def-use information when the execution of a regular instruction (not `athrow`) *always* results in an exception (This also affected the generation of the 3-address code; in the 3-address code this resulted in local variable usage which had no definition site.)
 - fixed the handling of exceptions by the data-flow framework if the exceptions hierarchy is not completely known
 - fixed an issue related to cyclic jumps; i.e. goto instructions which form a loop (notably x: goto x;)
 - fixed an infinite loop if a class file contains a "too small" unknown attribute and the reader (infinitely) waits for the next byte(s)
 - updated dependencies
 - removed support for setting the "isStrict" mode from AI (nowadays all VMs etc. perform all computations with asStrict semantics; we are now assuming everything is strict)

## 0.8.15 - Released Sep. 7th 2017
**General**
 - the call graph construction algorithms finally completely support Java 8 (e.g., default methods, static methods in interfaces, lambda expressions)
 - ***Assertions are turned-off by default when you checkout the latest stable release of OPAL***; to turn them on rename `local.sbt.template` to `local.sbt`; assertions are still turned on, when you depend on a development snapshot from Maven Central
 - removed the Eclipse plug-in sub-project; it wasn't maintained anymore and is now replaced by the ATOM plug-in
 - removed the Viz sub-project; it wasn't maintained anymore and OPAL has gained the possibility to generate SVGs using [a JavaScript based Graphviz version](https://github.com/mdaines/viz.js) which is executed using JDK's Nashorn JavaScript engine
 - fixed several minor bugs and issues when analyzing bytecode which contains compile-time dead code; the Groovy compiler frequently generated (generates?) such code
 - renamed packages called "analysis" to "analyses"

**Bytecode Representation Subproject**
 - `Method` and `Field` now have a back-link to their respective defining class files. The previous mechanisms provided by the `Project` are no longer available.
 - added explicit support for new code *entities* to better support Escape/Points-to analyses: `(Object/Array)AllocationSite` and `FormalParameter`
 - `ProjectInformationKey`s can now be initialized using project specific initialization information
   (See `Project.getOrCreateProjektInformationKeyInitializationData` for further information.)
 - renamed `ObjectImmutabilityAnalysi` to `ClassImmutabilityAnalysis` to reflect the names used in the forthcoming paper
 - removed the legacy method call resolution methods  (those only working with pre Java 8 code) from the `Project`
 - we are now *simplifying the control-flow at load-time* to ensure that the overall control-flow is more regular; i.e., that if instructions related to loops directly jump back and do not jump forward to a goto instruction which then jumps back
 - fixed a bug when INVOKEDYNAMIC instructions are rewritten

**Bytecode Disassembler**
 - significantly overhauled the HTML structure generated by the ***OPAL Disassembler***; take a look at our new ATOM Bytecode Disassembler plug-in!

**Abstract Interpretation Framework**
 - significantly improved the output generated by the `TAC` tool; the CFG is also included
 - Removed `FailingExpr` - to check if an expression has failed use the CFG (this was previously necessary anyway to check if a statement resulted in an exception; i.e., handling of exceptions is now more consistent.)
 - renamed JumpToSubroutine in the 3-address code to JSR
 - the parameters are now explicitly represented in TACAI
 - fixed a bug in TACAI where the handled exceptions are swallowed
 - fixed issues related to the computation of def-use information for very old (JSR/RET) bytecode
 - `Stmt` and `Expr` can now be cast using the "as..." methods, if the target type is known (this provides an alternative if a pattern match is only done w.r.t. a (small) subset of the declared fields.)

**Developer Tools**
 - ***Hermes*** now has a headless mode (HermesCLI) (*It is not yet possible to run the optimizer, but everything else works.*)

**Common**
 - renamed the methods of the `PropertyStore` which used operators (e.g., '<<', '<|<', ... ) to facilitate comprehension
 - clients can now control whether cycle-resolution should be performed or not
 - IntSet is now called IntArraySet

## 0.8.14
 - OPAL now has a very advanced 3-address code representation; go to www.opal-project.de to read about the 3-address code
 - fixed all known bugs when rewriting INVOKEDYNAMIC instruction (contributed by Andreas Muttscheller)
 - improved the public API of the `Abstract Interpretation Framework` to facilitate analyses using the 3-address representation

src/main/de/tud/cs/st/bat/CFG.pl
/*
  License (BSD Style License):
  Copyright (c) 2009, 2011
  Software Technology Group
  Department of Computer Science
  Technische Universität Darmstadt
  All rights reserved.   
 
  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

  - Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  - Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution. 
  - Neither the name of the Software Technology Group or Technische 
    Universität Darmstadt nor the names of its contributors may be used to 
    endorse or promote products derived from this software without specific 
    prior written permission. 

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  POSSIBILITY OF SUCH DAMAGE.
*/

% This library calculates the Control-flow Graph (CFG) of a method. 
%
% Version $Date: 2010-01-29 10:56:45 +0100 (Fri, 29 Jan 2010) $ $Rev: 1085 $
% Author Michael Eichberg (www.michael-eichberg.de)

% Documentation of the association list implementation:
% http://gollem.science.uva.nl/SWI-Prolog/Manual/assoc.html

% between(Start,End,Value) :-
%	Value is an (integer) value between Start and End (inclusive).
%[Built-in]between(Start,End,Start) :- Start <= End.
%[Built-in]between(Start,End,Value) :- Start < End,NextValue is Start + 1, between(NextValue,End,Value).

% TODO To improve the performance we should inline the calculation of the CFG instead of performing it in three steps. In particular the step where we calculate the successor BBs of all BBs is currently highly inefficient!


% next_instr(+MID,+CurrentPC,NextPC) :- 
%	determines the PC of the next instruction (if any!)
next_pc(MID,CurrentPC,NextPC) :-
 	NextPC is CurrentPC + 1,
	instr(MID,NextPC,_). % to make sure that there is another instruction!
	
	
% Future version, where instruction meta-information is explicitly available.
% E.g. given the following meta information:
% 	control_transfer(if_icmpne,conditionally)).
% Further, let's assume that we have the following instruction:
% 	instr(m_3,3,if_icmpne(12))
% Now, to determine, if an instruction starts a basic block, we can use the 
% following query:
%  bb_start_instr(MID,0) :- instr(MID,0,_). % the first instruction starts a basic block if the method has an implementation at all.
%	bb_start_instr(MID,PC) :- 
% 		instr(MID,CurrentPC,Instr), 
%		Instr =.. [Mnemonic|_],control_transfer(Mnemonic,T), T \= 'no',	
%		(	PC is CurrentPC + 1, instr(MID,PC,_); % to make sure that the next instruction exists
%			(	T = 'conditionally',
%				Instr =.. [_,Branchoffset|_],
%				PC is CurrentPC + BranchOffset
%			)
%		)
	

% bb_start_instr(MID,?PC) :-
% 	PC is the program counter of an instruction that starts a basic block.
bb_start_instr(MID,0) :- instr(MID,0,_). % the very first instruction - if the method has an implementation.
bb_start_instr(MID,PC) :- % every "start", "end" and "handler" instruction of a try - catch block
	method_exceptions_table(MID,Handlers),
	member(handler(StartPC,EndPC,HandlerPC,_),Handlers),
	(
		PC = StartPC;
		next_pc(MID,EndPC,PC);
		PC = HandlerPC
	). 
bb_start_instr(MID,PC) :- % the jump target of a conditional jump instructions and the instruction immediately following a conditional jump instruction
	conditional_jump(MID,CurrentPC,BranchOffset),
	(
		PC is CurrentPC + BranchOffset;
		PC is CurrentPC + 1 % if the code is correct this instruction has to exist
	).
bb_start_instr(MID,PC) :- % the jump target of unconditional jump instructions and the instruction immediately following a jump instruction ...
	jump(MID,CurrentPC,BranchOffset),
	(
		PC is CurrentPC + BranchOffset;
		next_pc(MID,CurrentPC,PC) % ... if it exists.
	).
bb_start_instr(MID,PC) :- % the instruction immediately following an athrow or instruction...
	(
		instr(MID,CurrentPC,athrow);
		return(MID,CurrentPC)
	),
	next_pc(MID,CurrentPC,PC). % ... if it exists.



% bb_end_instr(MID,PC) :-
% 	PC is the program counter of the last instruction of a basic block
%	of the method identified by MID.
bb_end_instr(MID,PC) :- 
	method_exceptions_table(MID,Handlers),
	member(handler(StartPC,EndPC,HandlerPC,_),Handlers),
	(
		(StartPC > 0, PC is StartPC-1);
		PC = EndPC ;
		(HandlerPC > 0, PC is HandlerPC-1)
	). 
bb_end_instr(MID,PC) :- % a control transfer instruction ends a basic block
	conditional_jump(MID,PC,_);
	jump(MID,PC,_);
	instr(MID,PC,athrow);
	return(MID,PC).
bb_end_instr(MID,PC) :- % the instruction before a jump target ends a basic block
	(
		conditional_jump(MID,CurrentPC,BranchOffset);
		jump(MID,CurrentPC,BranchOffset)
	),
	TargetPC is CurrentPC + BranchOffset,
	TargetPC > 0, 
	PC is TargetPC - 1.


% bbs(MID,BBs) :-
%	BBs is the list of all basic blocks of the method MID. A basic block
%	is a pair where the first value is the program counter of the first instruction
%	of the basic block and the second value is the last instruction of the 
%	basic block.
bbs(MID,BBs) :-
	setof(BBStartPC,bb_start_instr(MID,BBStartPC),BBStartPCs),
	setof(BBEndPC,bb_end_instr(MID,BBEndPC),BBEndPCs),
	bbs(BBStartPCs,BBEndPCs,BBs).
bbs([BBStartPC|BBStartPCs],[BBEndPC|BBEndPCs],[(BBStartPC,BBEndPC)|BBs]) :-
	bbs(BBStartPCs,BBEndPCs,BBs).
bbs([],[],[]).	


/* VERY INEFFICIENT - WE NEED TO USE THE SET OF ALL BASIC BLOCKS INSTEAD OF 
	RECALCULATING IT EVERY TIME!
% bb_pred(MID,BB,Pred) :-
% 	- MID the method's unique ID
%  - BB is a basic block
%	- Pred is a basic block that might have been executed before this basic block
bb_pred(MID,(FirstInstrPC,_),Pred) :-
	bb(MID,Pred),Pred = (_,LastInstrPC),
	Offset is (FirstInstrPC - LastInstrPC),
	(
		conditional_jump(MID,LastInstrPC,Offset);
		(
			conditional_jump(MID,LastInstrPC,_),
			FirstInstrPC =:= LastInstrPC + 1
		);
		jump(MID,LastInstrPC,Offset);
		(
			LastInstrPC is FirstInstrPC - 1,
			no_control_transfer(MID,LastInstrPC) % required for exception handling	
		)
	).
bb_pred(MID,(FirstInstrPC,_),Pred) :-
	method_exceptions_table(MID,Handlers),
	member(handler(StartPC,EndPC,FirstInstrPC,_),Handlers),
	bb(MID,Pred),Pred = (AFirstInstrPC,_),
	StartPC =< AFirstInstrPC,
	AFirstInstrPC =< EndPC.	
*/


% bb_succ(MID,BB,BBs,SuccBB) :-
%	MID is the ID of a method
% 	BB is a basic block of the method MID (a basic block is a pair (StartInstrPC,LastInstrPC))
%	BBs is the set of ALL basic blocks of the method MID
%	SuccBB is a basic block that may be executed directly after the given basic block BB
bb_succ(MID,(_,LastInstrPC),BBs,SuccBB) :-
	member(SuccBB,BBs),SuccBB = (FirstInstrPC,_),
	Offset is (FirstInstrPC - LastInstrPC),
	(
		jump(MID,LastInstrPC,Offset);
		conditional_jump(MID,LastInstrPC,Offset);
		( 	% fall-through case
			FirstInstrPC =:= LastInstrPC+1, 
			(
				conditional_jump(MID,LastInstrPC,_); 
				no_control_transfer(MID,LastInstrPC)
			)
		)
	).	
bb_succ_ex(MID,(FirstInstrPC,LastInstrPC),BBs,SuccExBB) :-	
	method_exceptions_table(MID,Handlers),
	member(handler(StartPC,EndPC,HandlerPC,_),Handlers),
	FirstInstrPC >= StartPC,
	LastInstrPC =< EndPC,
	SuccExBB = (HandlerPC,_),
	member(SuccExBB,BBs).

% cfg(MID,CFG) :-
%	CFG encodes the method's CFG. CFG is a list of pairs where the first value
%	of each pair is a basic block BB of the CFG and the second value is the list
%	of all successor basic blocks of BB. The list can contain the basic block BB.
%	E.G. CFG = [ ((0, 21), [ (22, 24)]), ((22, 24), [ (25, 31), (120, 127)]), ((..., ...), [...])|...].
cfg(MID,CFG) :-
	findall(
		(BB,SuccBBs),
		(	bbs(MID,BBs),
			member(BB,BBs),
			findall(
				SuccBB,
				(bb_succ(MID,BB,BBs,SuccBB);bb_succ_ex(MID,BB,BBs,SuccBB)),
				SuccBBs)
		),
		CFG
	).


df_ordering(CFG,BBsInDFO) :-
	RootNode=((0,_),_),
	member(RootNode,CFG),
	df_ordering([RootNode],CFG,[],BBsInDFO).

% df_ordering(CurrentNodes,CFG,BBsInDFOAccu,BBsInDFO) :-
%	- CurrentNodes is the list of nodes that need to be visited. Initially,
%		this is the list containing just the root node of the CFG.
%		A node is a pair (BB,SuccBBs) where the first value is a basic block and the
%		the second value is a list of basic blocks that may be executed after
%		the basic block identified using the first value.
%	- CFG a CFG as calculated by the cfg predicate.
%	- BBsInDFOAccu are the BBs that are already visited. Initially, this has to
%		be the empty set [].
%	- BBsInDFO are all BBs of the given CFG in depth first order (this is the output argument).
df_ordering([(BB,SuccBBs)|OtherNodes],CFG,BBsInDFOAccu,BBsInDFO) :-
	\+member(BB,BBsInDFOAccu),!,%green cut
	append(BBsInDFOAccu,[BB],NewBBsInDFOAccu),
	findall(
		SuccNode,
		(member(SuccBB,SuccBBs),member(SuccNode,CFG),SuccNode=(SuccBB,_)),
		SuccNodes
	),
	append(SuccNodes,OtherNodes,CurrentNodes),
	df_ordering(CurrentNodes,CFG,NewBBsInDFOAccu,BBsInDFO).
df_ordering([(BB,_)|OtherNodes],CFG,BBsInDFOAccu,BBsInDFO) :-
	member(BB,BBsInDFOAccu), !,%green cut
	df_ordering(OtherNodes,CFG,BBsInDFOAccu,BBsInDFO).
df_ordering([],_,BBsInDFO,BBsInDFO).
	
	
/*** 
 A node D dominates another node N (D dom N), if ervery path from the initial
 node of the flow graph to N goes throug D. Hence, every node dominates itself,
 and the entry node of a loop dominates all nodes in the loop.
 The Dominator relationship is usually represented using a dominator tree where
 only the immediate dominator realtionship is explicitly represented. 
***/
%dom(D_BB,N_BB,CFG) :-

	
	

% Some example queries:
%	1. For each basic block BB of a method (m_6) get all successor basic blocks.
% 		bbs(m_6,BBs),member(BB,BBs),findall(SuccBB,bb_succ(m_6,BB,BBs,SuccBB),SuccBBs).	
%	2. Get a set of all basic blocks where each basic block is associated with the 
%		list of successor basic blocks (and time the execution.)
% 		time(
%			findall(
%				(BB,SuccBBs),
%				(	bbs(m_6,BBs),
%					member(BB,BBs),
%					findall(SuccBB,bb_succ(m_6,BB,BBs,SuccBB),SuccBBs)
%				),
%				Nodes
%			)
%		).



/*
	We need abstractions to answer queries such as: 
	"Is <InputStream>.close() called on all paths?"

	Note, some methods do not terminate! E.g.,
	public void endless() {
		while (true) {
			System.out.println("endless!");
		}
	}

	public void endless();
  	Code:
   Stack=2, Locals=1, Args_size=1
   	0:	getstatic	#15; //Field java/lang/System.out:Ljava/io/PrintStream;
   	3:	ldc	#21; //String endless!
		5:	invokevirtual	#23; //Method java/io/PrintStream.println:(Ljava/lang/String;)V
		8:	goto	0
	LineNumberTable: 
		line 9: 0
		line 8: 8
	does not terminate!
*/


	

	